lib/molinillo/resolution.rb

```diff
@@ -241,82 +259,78 @@ def initial_state
-        # index, allowing us to short-circuit subsequent checks
+        # Get all the possible unwinds for the current conflict
         current_conflict = conflicts[name]
         binding_requirements = binding_requirements_for_conflict(current_conflict)
         unwind_details = unwind_details_for_requirements(binding_requirements)
-        return unwind_details if unwind_details.state_index == states.size - 2
 
-        # TODO: For now we're filtering out primary unwinds for performance reasons. Once we move
-        # to smart selection of previous unwinds we can remove this line.
-        add_to_previous_unwinds = unwind_details.state_index > -1 && unwind_details.relationship != :primary
+        # Look for a past conflict that could be unwound further / differently,
+        # and where doing so would affect the requirement tree for the current
+        # conflict
+        # TODO: Can we filter this further? Check there's a chance of resolution
+        # for primary / parent elements of the tree?
+        alternative = unused_unwind_options.select do |option_set|
+          next false unless option_set.first > unwind_details.first
+          (option_set.map(&:requirement).flatten & unwind_details.map(&:requirement).flatten).any?
+        end.sort_by(&:first).last
 
-        # Process previous conflicts
-        previous_unwinds.each do |unwind|
-          unwind_details = unwind_details_for_requirements(unwind.conflicting_requirements, unwind_details)
-        end
+        # Add the current unwind options to the `unused_unwind_options` array.
+        # (They'll filtered when we unwind, so no need to worry which option is
+        # being used, if any.)
+        state.unused_unwind_options << unwind_details
 
-        unwind_details.new_conflict_requirements = binding_requirements if add_to_previous_unwinds
-        unwind_details
+        # Return the alternative (by construction greater a higher index than
+        # the unwind details for this conflict, if present), or the original
+        # details
+        alternative && alternative.first || unwind_details.first
       end
 
       # @param [Conflict] conflict to be unwound from
       # @param [UnwindDetails] details of the currently proposed unwind details
       # @return [UnwindDetails] Details of the nearest index to which we could unwind to
       #    resolve the given conflict conflict
-      # rubocop:disable Metrics/CyclomaticComplexity
-      def unwind_details_for_requirements(binding_requirements, existing_unwind_details = nil)
-        maximal_index = states.size - 2
-        unwind_details = existing_unwind_details || UnwindDetails.new(-1, nil, nil, [])
+      def unwind_details_for_requirements(binding_requirements)
+        unwind_details = []
 
         binding_requirements.reverse_each do |r|
+          unwind_details << UnwindDetails.new(-1, requirement_tree_for(r), nil, binding_requirements)
```

Comment:
I think calculating the full requirement tree is rather slow... is it really necessary?